### ISSUE TITLE
(C#) Generic types cannot be exported

### ISSUE THREAD

**Issue by markdibarry** [godotengine/godot#74960]
### Godot version

4.1.dev (550a779)

### System information

Windows 10

### Issue description

An error is shown that Godot doesn't support the type even when it's constrained to a supported type:

![image](https://user-images.githubusercontent.com/21325943/225443873-7a625057-ef24-4205-92e6-e7d74a23b230.png)

### Steps to reproduce

Create an export property of a generic type constrained to a supported type.

### Minimal reproduction project

N/A

**Comment by Chaosus** [comment]
I don't think that it will be ever possible to fix, since Godot has more strict type system than C#. cc @raulsntos - maybe you explain it better.

**Comment by raulsntos** [comment]
You are not able to attach generic classes to nodes, so you can't attach `Foo<T>` to a node. However, you should be able to inherit from a generic class and attach that:

```csharp
public partial class Foo<T> : Node { }

public partial class FooDerived : Foo<Node> { }
```

Attaching `Foo<T>` wouldn't be possible, but attaching `FooDerived` should work. And since, at compile time, C# already knows that `T` is `Node` in this case, which is Variant-compatible, it seems like it should work.

The **GD0102** error comes from our source generators, we check to make sure you are not exporting an invalid type because otherwise it will fail during runtime so it's preferable to fail early.

I think the generator should be able to check the constraints of a generic type parameter to check if it's Variant-compatible and then avoid emitting the diagnostic.

Also, using the `[MustBeVariant]` attribute should probably make this use case work as well, but I'm not sure if it's currently supported.

```csharp
public partial class Foo<[MustBeVariant] T> : Node
{
    [Export]
    public T Bar { get; set; }
}
```

**Comment by markdibarry** [comment]
@raulsntos Unfortunately it doesn't (also without the constraint), but that's promising that it should be possible.
![image](https://user-images.githubusercontent.com/21325943/227728290-c115e599-b5a7-4829-ba7a-529da5c639b2.png)

**Comment by Shadowblitz16** [comment]
> I don't think that it will be ever possible to fix, since Godot has more strict type system than C#. cc @raulsntos - maybe you explain it better.

I was actually thinking godot has a much more relaxed typing system
But I agree with markdibarry this should be supported

**Comment by nrgill28** [comment]
For anyone following this issue, I'm working on adding support for this. You can track my progress over on my branch: https://github.com/nrgill28/godot/tree/csharp-generic-exports.

So far it mostly works for exported properties and it looks exactly as you'd expect. All Variant types are supported, so Resources, Nodes, ints, floats, etc. You can even do silly things like make the whole generic type itself an array of strings (example below).

Aside from exported properties, this also needs to work with script methods and signals and I haven't gotten around to that yet. There's also some concerns about some of the logic that previously could be done entirely at compile time now necessarily being done at runtime, but hopefully that will get sorted and we will have a fancy new feature to play with soon. 😄

Example code:
```cs
public partial class GenericResource<[MustBeVariant] T> : Resource
{
    [Export] public T Foo { get; set; }
}

[GlobalClass]
public partial class NonGenericResource : GenericResource<Godot.Collections.Array<string>>
{
    [Export] public int Bar;
}
```

And of course that looks as you'd expect in the inspector.
![image](https://github.com/godotengine/godot/assets/22647801/0ec4216f-a834-4b0d-b572-d5224e95c79d)

**Comment by nrgill28** [comment]
Update: got everything working, exported properties, fields, signals, and methods using generic type parameters are all accessible in the editor (and from GDScript). I do however have some things I want to get a bit of discussion on.

The first topic is what the desired behavior should be when someone tries to export a property (or signal, or method) of type `T[]`. [I've uploaded a code example to Gist to demonstrate](https://gist.github.com/nrgill28/4cbe7412011761424a30026f5415e907). The problem here is that while `T` is constrained to only variant types thanks to the `MustBeVariant` attribute, **not every variant type `T` has a valid `T[]`**. When exported as an array, types like `int`, `Vector2`, `Color`, and some others are marshaled as their packed array counterparts while the remaining variant types need to use the Godot array type instead. This currently is not a problem as without generic type support it is not possible to split the element type apart from the array declaration and the analyzer can always tell if it is valid from just the declaration alone, but with type parameters this is no longer the case.

In my current implementation I've decided to preserve the existing export behavior at all costs, exporting a `T[]` when `T` is `int` will result in the field being marshaled as a `PackedInt32Array` just as if you had declared it non-generically. To prevent situations where it may be possible for the generic class to export a non-variant type I have updated the analyzer to find usages of the generic classes and determine, given the type arguments it was used with, if any of its members would then result in these non variant types. Another far simpler solution might be to just disallow the use of `T[]` at the cost of then no longer being able to use the packed array types.

There are also two portions of my current implementation that may have drawbacks, both of which pertain to generating the hint strings for properties. With my current implementation I don't believe it's possible to generate hint strings without reflection, as it is needed in multiple places for things like obtaining enum type names and values, nested array hints, or just obtaining the class name of a custom Resource type. This means then that I am unsure if this code would be 100% safe in an AOT-compiled environment, however if the hint strings are only ever needed in the editor then that should be OK. The other drawback is since hint strings need to be generated at runtime, the logic to do that has been duplicated outside of the source generator and into the runtime library. This is the only place that has required duplicated logic.

As for possible other implementations should this one have problems, another idea I had would be to defer the source generation of generic members from the class in which they are declared into the class in which they have had their type completely specified. This has the advantage of being done entirely at compile time, however it would result in those exported members technically not belonging to the type they were declared in which would likely look odd in the editor (properties showing up under the wrong categories) and might have other side effects. I went with my current implementation over this other idea because I wanted to preserve as much of the existing behavior as possible.
